### ISSUE TITLE
add verifiable cryptographic signature to email event in "security history" log

### ISSUE THREAD

**Issue by monperrus** [pypi/warehouse#15974]
**What's the problem this feature will solve?**
Verifying the event trail in security history

**Describe the solution you'd like**
Add a field `signature` to the event
```
From: noreply@pypi.org
To: martin.monperrus@gnieh.org
Subject: [PyPI] Two-factor method added
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/simple;
	s=qnyyfqodjxtnsobvnrsglem3olqxgb5o; d=pypi.org; t=1716013242;
	h=Subject:From:To:MIME-Version:Content-Type:Message-ID:Date;
	bh=sBC+yG9GSD1CnHoXXbcy9jcoanUnWeFQgXevQ7vrQ/o=;
	b=Idqy0AQzQeKAKPCkC/MvBx68S+5bGEsHdJ+MQHRMLtuVYQ4oqG7CH3dvB4VPeB+/
	KUjxrOU/ew3veyGZLNk8igmcxcXgJLplKWxVwi4kpni8/u9FurWd7KWJsBo6UH9DO1H
	lp8nQdB1rqPq3dOFXhcsvNM0lt3X+H86bv6bXNjI=
```

**Additional context**
Any signature would work. The advantage of `DKIM-Signature` is that 1) it is already there 2) there is [email client support](https://github.com/lieser/dkim_verifier/) to verify DKIM-Signature

**Comment by woodruffw** [comment]
Making sure I understand: are you asking for a DKIM signature for the per-project "security history" page on `pypi.org` itself, or are you asking about the DKIM signature on the security emails themselves?

(If it's the latter, I think PyPI already issued DKIM-signed message bodies, since all of the emails in my inbox have the `DKIM-Signature` header. If it's the former, I think the value of signatures in the "security history" log would be minimal, since PyPI itself has transport security. But I could be misunderstanding.)

**Comment by monperrus** [comment]
I'm asking about  the DKIM signature on the security emails themselves.

> If it's the latter, I think PyPI already issued DKIM-signed message bodies

Yes, that's correct, and that's great.

The point here is to be able to cross check on two diverse channels (web + email) the security audit trail. Cross-checking with cryptographic signatures (DKIM in this case) is a good way to mitigate attacks on one channel.

**Comment by woodruffw** [comment]
> The point here is to be able to cross check on two diverse channels (web + email) for the security audit trail. Cross-checking with cryptographic signatures (DKIM in this case) is a good way to mitigate attacks on one channel.
> 

Could you provide an example of the kind of attack you have in mind? PyPI already has full transport security via TLS, so the audit trail you see on the website is signed during delivery. The two  DKIM/TLS signatures aren't comparable, but they wouldn't be in any case (since the website doesn't use the same layout for security events as the security emails do).

TL;DR: I'm not sure there's a security advantage to using the same DKIM key to sign PyPI's web-rendered security events. In practice, it might even result in a worse security posture: DKIM is currently encapsulated by PyPI's email provider, whereas doing this would require PyPI's admins to maintain a PKI for DKIM keys.

**Comment by monperrus** [comment]
> Could you provide an example of the kind of attack you have in mind?

The PyPI email provider is compromised, I receive an email "Change your password" with valid DKIM, pointing to a phishing website.

Mitigation: I verify the authenticity of the email on the website.

Note that this does not have to be the DKIM signature. It could be any kind of high-integrity cryptographic signature that would be present on both sides (email + web) for verification.

**Comment by woodruffw** [comment]
> Mitigation: I verify the authenticity of the email on the website.

This mitigation exists: when you connect to pypi.org, you're connecting via TLS. The security events that you can confirm on the relevant user/project page are served over an authenticated, encrypted connection (one that your user-agent negotiates for you).

Given that, I don't think there's significant added value to serving another set of signatures (and maintaining an associated PKI) over an already authenticated connection. Doing so is unlikely to benefit the majority of users (who don't verify signatures unless they're made mandatory) while imposing a substantial operational/maintenance burden (w/r/t maintaining and securing a new PKI).

**Comment by dstufft** [comment]
To rephrase what I think the proposed attack is here:

1. Email provider ends up compromised, giving the attack the chance to send emails using valid DKIM keys (and SPF records too).
2. PyPI sends an email, using the compromised email provider, registering an event in the event audit log.
3. Compromised email provider takes the valid email sent from PyPI, and modifies the content prior to sending it to the user (perhaps to inject phishing links or something).
4. User attempts to verify that the email is legit using DKIM and the security log on PyPI, which doesn't contain enough information to verify that it is actually the email that PyPI sent, just that it was sent from the correct provider.

It's not really a meaningfully different attack then what if the CDN provider was compromised, since in both cases the CDN provider or the email provider is responsible for validating that the content is from PyPI itself, and end users don't have an E2E means to verify that, they can only trust that our CDN or email provider isn't compromised and is providing them the correct content.

That being said, I agree with @woodruffw that I don't think there's enough value here to add the additional complication of trying to fix a compromised email sender.

My rationale for this is basically:

- We don't treat email as a secure channel, users can already get phishing emails that purport to be from us to varying degrees of authenticity. While a compromised email provider would effectively look authentic, the links in the email would not appear to be authentic, and a user that is being fastidious enough to manually verify the veracity of the email should be able to notice the domain is not one of our valid domains.
- It relies on human beings manually checking the veracity of the email, which is something that I suspect only a handful of people in the entire world would be willing to do on a regular basis.
- We already have better solutions for phishing resistance, through requiring 2FA and in particular, trying to guide people towards phishing resistant forms of MFA such as u2f tokens.
- Our email provider is AWS SES, and if they're compromised PyPI is not going to be the most attractive target on the list of things that the attacker has access to.
- Email content is often treated as malleable by clients, there's a long history of gpg signatures of mail ultimately ending up failing signature verification because a mail client decided to interpret the contents of the message and mutate it locally. DKIM doesn't suffer from this problem since typically it's the MTAs that are verifying DKIM not the mail clients.

So for all of that, this attack falls into a category where the attack itself is unlikely to happen, already has mitigations that are better than the proposed mitigation, and the proposed mitigation is fragile and requires a non trivial amount of effort both for users and maintenance for PyPI itself.

So I'm going to close this issue, but thanks for taking the time to open this issue!

**Comment by monperrus** [comment]
Agree on the trade-off, thanks a lot for the wrap up.

I must say that the "security history" log in PyPI is great, and [we](https://chains.proj.kth.se/) were super happy to find it.

Along the same line of thinking (package ecosystem integrity with cryptographic signatures), would you be interested if we open and elaborate on an issue "add transparency log on top of the security history trail"?
